### PR TITLE
Loose Tokens

### PR DESCRIPTION
**Pull request by privat**
Another step to simplify the clients of AST: loose tokens.

Some tokens, correctly lexed by the lexer, are discarded by the parser because of (thanks to?) sablecc transformations. Mainly comments and new-lines (EOL) are lost but some other tokens might be also discarded (currently, commas and dots are still lost).

Previously, the full sequence of tokens can still be accessed trough `next_token` and `prev_token` but this mean that tools like nitprettty and nitlight have to maintain two cursors, one on the AST and one on this linked list of tokens and try to keep these two cursors synchronized.

This PR names the concept of *loose* tokens that are lexed tokens but absent from the AST, and attach to each (non-loose) token in the AST a list of loose tokens that precede it (`prev_looses`) and that follow it (`next_looses`).
This way, a visit on the AST can be used to also easily process all the tokens from the source.

Future PRs will try to use this new thing to improve/simplify/whatever nitpretty and nitlight.

### CODE REVIEW

File: src/parser/parser_work.nit [hunk 1]
```diff
@@ -176,21 +177,48 @@ end
 # Uses existing token locations to infer location of productions.
 private class ComputeProdLocationVisitor
 	super Visitor
+	# The current (or starting) token used to collect precede/follow
```
Comment (Morriar) [inline src/parser/parser_work.nit]: Did you mean prev, next?

File: src/parser/parser_work.nit [hunk 2]
```diff
@@ -176,21 +177,48 @@ end
...
 	do
 		if n isa Token then
+			# Skip injected tokens
 			if not isset n._location then return
+
+			# Collect token not in the AST and attach them to token in the AST
+			var t = token
+			if n != t then
+				# in order, we have the tokens `lt` then `t` then `n`
+				# We will advance `t` to attach then to `lt` (follow) or to `n` (precede)
```
Comment (Morriar) [inline src/parser/parser_work.nit]: This seems important, but I don't understand.

File: src/parser/parser_nodes.nit [hunk 3]
```diff
@@ -311,6 +311,12 @@ abstract class Token
 	# May have disappeared in the AST
 	var next_token: nullable Token = null
 
+	# Loose tokens that precede `self`.
```
Comment (Morriar) [inline src/parser/parser_nodes.nit]: You should state the case where a loose token will be associated before.

File: src/parser/parser_nodes.nit [hunk 4]
```diff
@@ -311,6 +311,12 @@ abstract class Token
...
+	var prev_looses = new ANodes[Token](self)
+
+	# Loose tokens that follows `self`
```
Comment (Morriar) [inline src/parser/parser_nodes.nit]: You should state the case where a loose token will be associated after.

**Comment by Morriar** [comment]
Fix the documentation and you have my +1

**Comment by jpages** [comment]
+1

**Comment by R4PaSs** [comment]
:shipit:

**Comment by privat** [comment]
reroll: rewrite documentation and try to pass the tests

File: src/parser/parser_nodes.nit [hunk 5]
```diff
@@ -311,6 +311,33 @@ abstract class Token
 	# May have disappeared in the AST
 	var next_token: nullable Token = null
 
+	# Is `self` a token discarded from the AST?
+	#
+	# Loose tokens are not present in the AST.
+	# It means they are parsed by the lexer but where discarded by the parser.
```
Comment (Morriar) [inline src/parser/parser_nodes.nit]: were*

Comment (Morriar) [inline src/parser/parser_nodes.nit]: Actually, your sentence is both in the present and passive form, it bothers me.

Maybe use `but are discarded`

File: src/parser/parser_nodes.nit [hunk 6]
```diff
@@ -311,6 +311,33 @@ abstract class Token
...
+	# It also means that they are not visited of manipulated by AST-related functions.
```
Comment (Morriar) [inline src/parser/parser_nodes.nit]: or manipulated* ?

File: src/parser/parser_nodes.nit [hunk 7]
```diff
@@ -311,6 +311,33 @@ abstract class Token
...
+	#
+	# Each loose token is attached to the non-loose token that precedes or follows it.
+	# The rules are a following:
```
Comment (Morriar) [inline src/parser/parser_nodes.nit]: the following*

File: src/parser/parser_nodes.nit [hunk 8]
```diff
@@ -311,6 +311,33 @@ abstract class Token
...
+	#
+	# * tokens that follow a non-loose token on a same line are attached to it
```
Comment (Morriar) [inline src/parser/parser_nodes.nit]: is this `next_looses`?

File: src/parser/parser_nodes.nit [hunk 9]
```diff
@@ -311,6 +311,33 @@ abstract class Token
...
+	# * other tokens are attached to the non-loose token that follows it (on the same line or the next one)
```
Comment (Morriar) [inline src/parser/parser_nodes.nit]: is this something like `next.prev_looses`?

File: src/parser/parser_nodes.nit [hunk 10]
```diff
@@ -311,6 +311,33 @@ abstract class Token
...
+	#
+	# Loose tokens are mostly end of lines (`TEol`) and comments (`TComment`).
+	# Whitespace are ignored by the lexer, so they are not even considered as loose tokens.
+	# See `blank_before` to get the blank space that separate tokens.
```
Comment (Morriar) [inline src/parser/parser_nodes.nit]: spaces?

File: src/parser/parser_nodes.nit [hunk 11]
```diff
@@ -311,6 +311,33 @@ abstract class Token
...
+	var is_loose = false
+
+	# Loose tokens that precede `self`.
+	#
+	# These tokens thus start the line or belong to a line with only loose tokens.
```
Comment (Morriar) [inline src/parser/parser_nodes.nit]: remove the `thus`

File: src/parser/parser_nodes.nit [hunk 12]
```diff
@@ -311,6 +311,33 @@ abstract class Token
...
+	var prev_looses = new Array[Token] is lazy
+
+	# Loose tokens that follows `self`
```
Comment (Morriar) [inline src/parser/parser_nodes.nit]: follow*

**Comment by privat** [comment]
reroll: spell and improved doc a little more

**Comment by Morriar** [comment]
+1

**Pull request merged**